### ISSUE TITLE
Automatically open terminal when Stdout is not a terminal

### ISSUE THREAD

**Issue by wolfmagnate** [charmbracelet/bubbletea#860]
# **Is your feature request related to a problem? Please describe.**

## description
Currently, to redirect output results to another file or pipe them to another command is not supported. Therefore, if you write bubbles tea's standard output to a file with the current implementation, the file will have ANSI control characters written directly to it.
this problem is reported in #823.

## why is this a problem?
- TUI applications will be able to take full advantage of the CLI application, such as the convenience of pipes and redirection.
- It is strange that the input supports pipes and redirects [here in tea.go](https://github.com/charmbracelet/bubbletea/blob/5536bca34e218a1e17a55e5ba36c22fab5784f44/tea.go#L416C2-L452C3), but the output does not.

# **Describe the solution you'd like**

If the standard output of the bubble tea program is connected to something other than a terminal, the program will determine this at startup and change the destination of bubble tea's TUI to the terminal and the destination of stdout to the output specified at startup.

Add the following process
- check whether stdout is a terminal at the start of the program.
- if it is a terminal, proceed with the process as it is now. (just output to stdout)
- if it is not a terminal, open a new "dev/tty" and output to it.

# **Describe alternatives you've considered**
- You can use a golang program to write an output to a file, but it is inconvenient.
- open dev/tty and use `WithOutput` option
```go
func main() {
	tty, err := os.OpenFile("/dev/tty", os.O_WRONLY, 0)
	if err != nil {
		panic(err)
	}
	defer tty.Close()

	p := tea.NewProgram(initialModel(), tea.WithOutput(tty))
	fmt.Println("result of command")
	if _, err := p.Run(); err != nil {
		fmt.Println(err)
		os.Exit(1)
	}
}
```
This works on linux, but this is not crossplatform solution. Also, this is inconsistent interface because you don't have to write any code for input, but have to write for output.

# **Additional context**
I am not sure where i should write the statements to solve this problem.
Currently, Program.output is initialized [here](https://github.com/charmbracelet/bubbletea/blob/5536bca34e218a1e17a55e5ba36c22fab5784f44/tea.go#L203C2-L211C72). I should add the new process here in terms of extending the existing process.
[Here](https://github.com/charmbracelet/bubbletea/blob/5536bca34e218a1e17a55e5ba36c22fab5784f44/tea.go#L416C2-L452C3) is a program from tea.go that creates a new TTY for the input source. The input and output processing are similar in purpose and content, so I should add the processing here.

**Comment by wolfmagnate** [comment]
I would like to write code to solve this problem and create a pull request, but before I do so I would like to hear some design advice on how it should be implemented.

**Comment by curusarn** [comment]
Hi @wolfmagnate,

I'm trying to use the workaround you mentioned and it seems that color formatting doesn't work when I pass `/dev/tty` as output to `NewProgram()`.

I've reproduced the issue using two of examples in this repo.
I've ran the commands as `echo $(go run main.go)`.

For `examples/list-fancy/main.go` no colors are showing up:
![CleanShot 2024-05-04 at 18 17 14@2x](https://github.com/charmbracelet/bubbletea/assets/10132717/cb4c2a9b-019c-47ff-808a-ebd566b188f7)

For `examples/glamour/main.go` colors from glamour markdown auto-styling do work but normal lipstick styles still do not work - i.e. the border should be purple (see another screenshot below):
![CleanShot 2024-05-04 at 18 16 00@2x](https://github.com/charmbracelet/bubbletea/assets/10132717/e7abfee9-527c-4778-b5ab-14369e3e3242)

Outputs when running the commands normally as `go run main.go` for comparison:
![CleanShot 2024-05-04 at 18 16 45@2x](https://github.com/charmbracelet/bubbletea/assets/10132717/69dae985-039a-477b-822a-a86c94b77a08)
![CleanShot 2024-05-04 at 18 16 24@2x](https://github.com/charmbracelet/bubbletea/assets/10132717/56799eab-cca2-4cdf-aea4-b5244fcdc841)


Gist with full code I used for both examples:
- https://gist.github.com/curusarn/f9057ca6926ab23e1c497641e7f2728b#file-issue_demo_list_fancy-go
- https://gist.github.com/curusarn/f9057ca6926ab23e1c497641e7f2728b#file-issue_demo_glamour-go
     - Glamour styles that do work: https://gist.github.com/curusarn/f9057ca6926ab23e1c497641e7f2728b#file-issue_demo_glamour-go-L97
     - Lipstick styles that don't work - e.g.: https://gist.github.com/curusarn/f9057ca6926ab23e1c497641e7f2728b#file-issue_demo_glamour-go-L91

Do you have any pointers or advice on how to make all styles work? 🙏 
Sorry for hijacking this issue a bit - I'm happy to create a new one if that work better.

Many thanks!
Simon

**Comment by jmonroynieto** [comment]
It took me way too long to find this workaround. The first improvement may be to make this info a bit more accessible. Because of Gum, I  expected this to be an out-of-the-box behaviour.

Following the gum train of thought, I tried to model my program for it, but it's hard to follow since the call stack delves into Kong cli, which relies on reflection and other unsafe code. Learning the internals of a framework I won't even use seems like a step too far into procrastination so I am happy to find a workaround.

If maybe any of you know how Gum is doing it, please let me know in this [discussion thread](https://github.com/charmbracelet/gum/discussions/570).

**Comment by dsh2dsh** [comment]
Hi,

I got the same issue. I'm creating 
```go
tea.NewProgram(model, tea.WithOutput(os.Stderr))
```
and I see all colors if I just run my program. But if I redirect the stdout like
```
go run . > /dev/null
```
everything is black and white, no colors.

**Comment by dsh2dsh** [comment]
Hi,

I found a workaround for my issue. My program uses [lipgloss](https://github.com/charmbracelet/lipgloss) and outputs to `os.Stderr` as I mentioned above. It works find, all output is colorful. But if I'll redirect `stdout` my program loose colors, as I mentioned above. I found that's because lipgloss by default uses `stdout` and when it redirected, lipgloss thinks the output isn't TTY or something like that. So for me the fix is

```go
lipgloss.SetDefaultRenderer(lipgloss.NewRenderer(os.Stderr))
```

before creating any `lipgloss.Style`.

Sorry for the noise, this isn't bubbletea's problem.

**Comment by dsh2dsh** [comment]
And one more fix for my issue. My program uses progress from bubbles and it also looses colors, with redirected `stdout`. That's because it does:

```
colorProfile:   termenv.ColorProfile(),
```

but termenv's default output is `stdout` and if `stdout` redirected, termenv thinks it isn't a TTY or something like that. To fix it I added

```go
termenv.SetDefaultOutput(termenv.NewOutput(os.Stderr))
```

before first using of progress.
